grpc/src/main/java/com/linecorp/armeria/common/grpc/protocol/AbstractUnaryGrpcService.java

```diff
@@ -0,0 +1,122 @@
+     * Returns an unframed response message to return to the client, given an unframed request message. It is
+     * expected that the implementation has the logic to know how to parse the request and serialize a response
+     * into {@code byte[]}. The returned {@code byte[]} will be framed and returned to the client.
+     */
+    protected abstract CompletableFuture<byte[]> handleMessage(byte[] message);
+
+    @Override
+    protected final HttpResponse doPost(ServiceRequestContext ctx, HttpRequest req) {
+        final CompletableFuture<HttpResponse> responseFuture =
+                req.aggregateWithPooledObjects(ctx.contextAwareEventLoop(), ctx.alloc())
+                   .thenCompose(msg -> deframeMessage(msg.content(), ctx.alloc()))
+                   .thenCompose(requestMessage -> {
+                       final byte[] requestMessageBytes;
+                       try {
+                           requestMessageBytes = ByteBufUtil.getBytes(requestMessage);
+                       } finally {
+                           requestMessage.release();
+                       }
+                       return handleMessage(requestMessageBytes);
+                   })
+                   .thenApply(responseMessage -> {
+                       final ArmeriaMessageFramer framer = new ArmeriaMessageFramer(
+                               ctx.alloc(), Integer.MAX_VALUE);
+                       final HttpData framed = framer.writePayload(Unpooled.wrappedBuffer(responseMessage));
+                       return HttpResponse.of(
+                               RESPONSE_HEADERS,
+                               framed,
+                               GrpcTrailersUtil.statusToTrailers(StatusCodes.OK, null, true));
+                   })
+                   .exceptionally(t -> {
+                       final HttpHeaders trailers;
+                       if (t instanceof ArmeriaStatusException) {
+                           ArmeriaStatusException statusException = (ArmeriaStatusException) t;
+                           trailers = GrpcTrailersUtil.statusToTrailers(
+                                   statusException.getCode(), statusException.getMessage(), false);
+                       } else {
+                           trailers = GrpcTrailersUtil.statusToTrailers(
+                                   StatusCodes.INTERNAL, t.getMessage(), false);
+                       }
+                       return HttpResponse.of(trailers);
+                   });
+
+        return HttpResponse.from(responseFuture);
+    }
+
+    private CompletableFuture<ByteBuf> deframeMessage(HttpData framed, ByteBufAllocator alloc) {
+        final CompletableFuture<ByteBuf> deframed = new CompletableFuture<>();
+        try (ArmeriaMessageDeframer deframer = new ArmeriaMessageDeframer(
+                new Listener() {
+                    @Override
+                    public void messageRead(ByteBufOrStream message) {
+                        // Compression not supported.
```

Comment:
question. does grpc compress by default?

Reply:
Nope, users have to explicitly enable it. And even if a client happened to enabled it, since we don't return any supported encodings in the response headers, they will fallback to sending uncompressed messages anyways.

Reply:
Nope, users have to explicitly enable it. And even if a client happened to enabled it, since we don't return any supported encodings in the response headers, they will fallback to sending uncompressed messages anyways.